arangod/Aql/SortedCollectExecutor.cpp

```diff
@@ -377,9 +388,181 @@ std::pair<ExecutionState, size_t> SortedCollectExecutor::expectedNumberOfRows(si
+
+    // TODO store in member if we have not been called with data before
+    if (state == ExecutorState::DONE && !(_haveSeenData || input.isInitialized())) {
+      // we have never been called with data
+      LOG_DEVEL_SC << "never called with data";
+      if (_infos.getGroupRegisters().empty()) {
+        // by definition we need to emit one collect row
+        _currentGroup.writeToOutput(output, InputAqlItemRow{CreateInvalidInputRowHint{}});
+        rowsProduces += 1;
+      }
+      break;
+    }
+
+    // either state != DONE or we have an input row
+    TRI_ASSERT(state == ExecutorState::HASMORE || state == ExecutorState::DONE);
+    if (!input.isInitialized() && state != ExecutorState::DONE) {
+      LOG_DEVEL_SC << "need more input rows";
+      break;
+    }
+
+    TRI_IF_FAILURE("SortedCollectBlock::getOrSkipSomeOuter") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+
+    TRI_IF_FAILURE("SortedCollectBlock::hasMore") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
+
+    if (input.isInitialized()) {
+      _haveSeenData = true;
+
+      // if we are in the same group, we need to add lines to the current group
+      if (_currentGroup.isSameGroup(input)) {
+        LOG_DEVEL_SC << "input is same group";
+        _currentGroup.addLine(input);
+
+      } else if (_currentGroup.isValid()) {
+        LOG_DEVEL_SC << "input is new group, writing old group";
+        // Write the current group.
+        // Start a new group from input
+        rowsProduces += 1;
+        _currentGroup.writeToOutput(output, input);
+
+        if (output.isFull()) {
+          LOG_DEVEL_SC << "now output is full, exit early";
+          pendingGroup = true;
+          _currentGroup.reset(InputAqlItemRow{CreateInvalidInputRowHint{}});
+          break;
+        }
+        _currentGroup.reset(input);  // reset and recreate new group
+
+      } else {
+        LOG_DEVEL_SC << "generating new group";
+        // old group was not valid, do not write it
+        _currentGroup.reset(input);  // reset and recreate new group
+      }
+    }
+
+    inputRange.nextDataRow();
+
+    bool produceMore = !output.isFull();
+    if (!produceMore) {
+      pendingGroup = true;
+      break;
+    }
+
+    if (state == ExecutorState::DONE) {
+      rowsProduces += 1;
+      _currentGroup.writeToOutput(output, input);
+      _currentGroup.reset(InputAqlItemRow{CreateInvalidInputRowHint{}});
+      break;
+    }
+  }
+
+  auto newState = pendingGroup ? ExecutorState::HASMORE : inputRange.upstreamState();
+
+  LOG_DEVEL_SC << "reporting state: " << newState;
+  return {newState, Stats{}, AqlCall{}};
+}
+
+auto SortedCollectExecutor::skipRowsRange(AqlItemBlockInputRange& inputRange, AqlCall& clientCall)
+    -> std::tuple<ExecutorState, size_t, AqlCall> {
+  TRI_IF_FAILURE("SortedCollectExecutor::skipRowsRange") {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+  }
+
+  TRI_ASSERT(clientCall.needSkipMore());
+  while (clientCall.needSkipMore()) {
```

Comment:
This is not a hard-stop for me.
However reading this review I think. skip and produce are almost identical.

There are only 2 parts that actually differ:
a) while (clientCall.needSkipMore()) vs. while(!output.full())
b) clientCall.didSkip(1) vs output.writeToOutput()


do you think it is somehow possible to make the other code-parts shared?
This would reduce the amount of possible bugs by 50% ;)